### ISSUE TITLE
Bug: Dynamic Convolution Attention fails in `mixed_precision` training.

### ISSUE THREAD

**Issue by erogol** [coqui-ai/TTS#417]
**Describe the bug**
Dynamic Convolutional Attention fails in mixed_precision training and ultimately causes NaN error. 

**To Reproduce**
1. set ```mixed_precision=True``` in ```config.json```.
2. set ```dynamic_convolution=True``` in ```config.json```.
3. start training a tacotron or tacotron2 model.
4. On TB initially you observe broken attention alignment.
5. Ultimately loss becomes NaN.

**Expected behavior**
The model should learn the alignment after 10K iterations with no NaN loss as it does in full precision training.

**Environment (please complete the following information):**
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
- **PyTorch or TensorFlow version (use command below)**: Torch 1.8.0
- **Python version**: 3.8
- **CUDA/cuDNN version**: 11.2
- **GPU model and memory**: 1080Ti
- **Exact command to reproduce**:

**Additional context**

**Comment by SanjaESC** [comment]
Question: Did you try disabling SSIM loss? 
When debugging with https://pytorch.org/docs/stable/autograd.html#anomaly-detection
SSIM was the culprit returning the NaN values (also see https://github.com/pytorch/pytorch/issues/40497#issuecomment-709846922).

**Comment by erogol** [comment]
In my case they all return NaN but I've not checked like you did. I should try without SSIM then.

**Comment by a-froghyar** [comment]
Is it working fine if mixed precision is off?

**Comment by erogol** [comment]
@a-froghyar yes it does

**Comment by a-froghyar** [comment]
@erogol In my experiments without mixed precision, DCA fails (returns NaNs) after almost precisely 3 epochs - did 3 trials on a model that normally runs fine with original attention and DCA always fails around the same time.

**Comment by a-froghyar** [comment]
And just in case this helps, I also got OOM errors when using DCA (same batch size as with original attention) and I was not maxing out my GPU

**Comment by erogol** [comment]
Probably until I experience an error myself I won't spend time on this issue. However, feel free to debug yourself.

It is normal for DCA to use more memory, but I don't know about "not maxing out the GPU". Currently, things run alright on my system if it helps; 

Ubuntu 20.04
Torch 1.8.1
python 3.8

**Comment by Sadam1195** [comment]
I am not sure if it helps but I was getting the same error I fixed the issue by setting `r = 6` `"gradual_training": [[0, 6, 64], [15000, 4, 64], [30000, 2, 32]] ` and setting `ddc_r = 6` removing all odd values of `r` and `ddc_r` helped my case but alignments were still on and off for the most of time during training.

**Comment by Sadam1195** [comment]
> Question: Did you try disabling SSIM loss?
> 
> When debugging with https://pytorch.org/docs/stable/autograd.html#anomaly-detection
> SSIM was the culprit returning the NaN values (also see [pytorch/pytorch#40497 (comment)](https://github.com/pytorch/pytorch/issues/40497#issuecomment-709846922)).

Have you checked this? https://github.com/pytorch/pytorch/issues/40497#issuecomment-764027496
@SanjaESC

**Comment by erogol** [comment]
I think this issue relates to https://github.com/pytorch/pytorch/issues/40497

**Comment by erogol** [comment]
using APEX backend with the new API seemingly helps
